pkg/controller/constrainttemplate/constrainttemplate_controller.go

```diff
@@ -79,7 +79,14 @@ func (a *Adder) InjectWatchManager(wm *watch.WatchManager) {
 
 // newReconciler returns a new reconcile.Reconciler
 func newReconciler(mgr manager.Manager, opa *opa.Client, wm *watch.WatchManager) (reconcile.Reconciler, error) {
-	constraintAdder := constraint.Adder{Opa: opa}
+	// constraintsCache stores total number of constraints per enforcement action
+	constraintsCache := map[string]map[string]bool{
+		"deny":         make(map[string]bool),
+		"dryrun":       make(map[string]bool),
+		"unrecognized": make(map[string]bool),
+	}
```

Comment:
Here and over in enforcement_action.go: from talking to @maxsmythe , my understanding is that "audit" is also a supported enforcement action. Also since enforcement actions are essentially an enum rather than a free string, consider defining a typed string to give yourself the compile-time safety of an enum, eg:
type EnforcementAction string
const (
  EnforcementAction deny = "deny",
  ...
)
var supportedEnforcementActions = []EnforcementAction{deny, dryrun, audit, unrecognized}

Reply:
supported enforcement actions are `deny` and `dryrun`, audit is not an enforcement action.
updated with the typed strings.